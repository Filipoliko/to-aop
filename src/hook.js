export function createHook(name, regular, callback) {
  return {
    [name]: function hook(meta) {
      return hookFor(meta, regular, callback);
    }
  };
}

export function hookFor(meta, filterRule, callback) {
  if (typeof filterRule === 'string') {
    if (meta.property.includes(filterRule)) {
      return callback(meta);
    }

    return null;
  }
  if (filterRule instanceof RegExp) {
    if (filterRule.test(meta.property)) {
      return callback(meta);
    }

    return null;
  }

  if (typeof filterRule === 'function') {
    if (filterRule(meta)) {
      return callback(meta);
    }

    return null;
  }

  if (Array.isArray(filterRule)) {
    return filterRule.map(({ rule, action }) => {
      return hookFor(meta, rule, action || callback);
    });
  }

  throw new TypeError(
    `Invalid rule type ${typeof filterRule}. Method accept string, regexp, function and array.`
  );
}

export const hookName = Object.freeze({
  beforeMethod: 'beforeMethod',
  afterMethod: 'afterMethod',
  aroundMethod: 'aroundMethod',
  beforeGetter: 'beforeGetter',
  afterGetter: 'afterGetter',
  aroundGetter: 'aroundGetter',
  beforeSetter: 'beforeSetter',
  afterSetter: 'afterSetter',
  aroundSetter: 'aroundSetter'
});
