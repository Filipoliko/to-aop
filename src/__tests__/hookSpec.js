import { createHook, hookFor } from '../hook';

describe('hook', () => {
  describe('method createHook', () => {
    it('should return specific hook structure', () => {
      const hook = createHook('beforeMethod', 'setMethod', () => {});

      expect(hook).toMatchSnapshot();
      expect(hook.beforeMethod.toString()).toMatchSnapshot();
    });
  });

  describe('method hookFor', () => {
    let metaTruthy = null;
    let metaFalsy = null;
    let fn = null;

    beforeEach(() => {
      metaTruthy = {
        property: 'setState'
      };
      metaFalsy = {
        property: 'getState'
      };
      fn = jest.fn();
    });

    it('should call action for filtering rule as string', () => {
      hookFor(metaTruthy, 'setState', fn);

      expect(fn).toHaveBeenCalledWith(metaTruthy);
    });

    it('should call action for filtering rule as regexp', () => {
      hookFor(metaTruthy, /setState/gi, fn);

      expect(fn).toHaveBeenCalledWith(metaTruthy);
    });

    it('should call action for filtering rule as function', () => {
      hookFor(metaTruthy, () => true, fn);

      expect(fn).toHaveBeenCalledWith(metaTruthy);
    });

    it('should call action hook for filtering rule as list', () => {
      hookFor(metaTruthy, [{ rule: 'setState', action: fn }]);

      expect(fn).toHaveBeenCalledWith(metaTruthy);
    });

    it('should throw error for bad structure at filtering rule', () => {
      expect(() => hookFor(metaTruthy, {}, fn)).toThrow();
    });

    it('should not call action', () => {
      hookFor(metaFalsy, () => false, fn);

      expect(fn).not.toHaveBeenCalled();
    });
  });
});
