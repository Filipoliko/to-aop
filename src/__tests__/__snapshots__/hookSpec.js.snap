// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hook method createHook should return specific hook structure 1`] = `
Object {
  "beforeMethod": [Function],
}
`;

exports[`hook method createHook should return specific hook structure 2`] = `
"function hook(meta) {
      return hookFor(meta, regular, callback);
    }"
`;
